notebook_utils/berdl_notebook_utils/spark/connect_server.py

```diff
@@ -0,0 +1,278 @@
+        self.spark_master_url = os.environ.get("SPARK_MASTER_URL", "spark://spark-master:7077")
+
+        # Process management
+        self.pid_file_path = Path(f"/tmp/spark-connect-server-{self.username}.pid")
+        self.log_file_path = self.connect_server_log_dir / f"spark-connect-server-{self.username}.log"
+
+    def create_directories(self) -> None:
+        """Create all required directories for Spark Connect server."""
+        self.user_conf_dir.mkdir(parents=True, exist_ok=True)
+        self.spark_event_log_dir.mkdir(parents=True, exist_ok=True)
+        self.connect_server_log_dir.mkdir(parents=True, exist_ok=True)
+
+    def generate_spark_config(self) -> None:
+        """Generate spark-defaults.conf with user-specific configurations."""
+        if not Path(self.template_path).exists():
+            raise FileNotFoundError(f"Spark config template not found: {self.template_path}")
+
+        # Copy base template
+        shutil.copy(self.template_path, self.spark_defaults_path)
+        logger.info(f"Copied base config from {self.template_path}")
+
+        # Append dynamic user-specific configurations
+        with open(self.spark_defaults_path, "a") as f:
+            f.write("\n# Dynamic user-specific configurations\n")
+            f.write(f"# Generated for user: {self.username}\n\n")
+
+            # Spark event log directory (for Spark History Server)
+            f.write(f"spark.eventLog.dir={self.spark_event_log_dir}\n")
+
+            # Hive metastore URI
+            f.write(f"spark.hadoop.hive.metastore.uris={self.settings.BERDL_HIVE_METASTORE_URI}\n")
+
+            # MinIO S3 configuration with user credentials
+            f.write(f"spark.hadoop.fs.s3a.endpoint={self.settings.MINIO_ENDPOINT_URL}\n")
+            f.write(f"spark.hadoop.fs.s3a.access.key={self.settings.MINIO_ACCESS_KEY}\n")
+            f.write(f"spark.hadoop.fs.s3a.secret.key={self.settings.MINIO_SECRET_KEY}\n")
+
+        logger.info(f"Spark configuration written to {self.spark_defaults_path}")
+
+
+class SparkConnectServerManager:
+    """Manager for Spark Connect server lifecycle."""
+
+    def __init__(self, config: Optional[SparkConnectServerConfig] = None):
+        """
+        Initialize Spark Connect server manager.
+
+        Args:
+            config: SparkConnectServerConfig instance. If None, creates default config.
+        """
+        self.config = config or SparkConnectServerConfig()
+
+    def is_running(self) -> bool:
+        """
+        Check if Spark Connect server is already running.
+
+        Returns:
+            True if server is running, False otherwise.
+        """
+        if not self.config.pid_file_path.exists():
+            return False
+
+        try:
+            with open(self.config.pid_file_path, "r") as f:
+                pid = int(f.read().strip())
+            # Check if process exists (doesn't actually kill it)
+            os.kill(pid, 0)
```

Comment:
Interesting, I would have used psutil instead of os.kill. I didn't know you could do that.